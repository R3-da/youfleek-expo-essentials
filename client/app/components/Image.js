import React, { useContext } from "react";
import { Image as CacheImage } from "react-native-expo-image-cache";
import { Image } from "react-native";
import { ThemeContext } from "../contexts/theme-context";

export default function CustomImage({
  src = null,
  imageUri = null,
  preview = null,
  cache = false,
  rounded = true,
  style = {},
  ...props
}) {
  const { theme } = useContext(ThemeContext);

  let defaultStyles;
  if (rounded) {
    defaultStyles = { width: "100%", height: "100%", borderRadius: 10 };
  } else {
    defaultStyles = { width: "100%", height: "100%" };
  }

  if (cache) {
    return (
      <CacheImage
        testID="cached-image"
        uri={imageUri}
        preview={{ uri: preview }}
        tint={theme}
        style={style ? { ...defaultStyles, ...style } : defaultStyles}
        {...props}
      />
    );
  }

  if (imageUri) {
    return (
      <Image
        testID="uri-image"
        source={{ uri: imageUri }}
        style={style ? { ...defaultStyles, ...style } : defaultStyles}
        {...props}
      />
    );
  }

  return (
    <Image
      testID="local-image"
      source={src}
      style={style ? { ...defaultStyles, ...style } : defaultStyles}
      {...props}
    />
  );
}
