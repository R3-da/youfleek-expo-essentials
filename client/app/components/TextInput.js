//
import React from "react";
import {
  View,
  StyleSheet,
  TouchableOpacity,
  TouchableWithoutFeedback,
} from "react-native";

import { Icon, Input } from "@ui-kitten/components";
import { MaterialIcons } from "@expo/vector-icons";

import Caption from "./Caption";

const styles = StyleSheet.create({
  captionContainer: {
    display: "flex",
    flexDirection: "row",
    alignItems: "center",
    marginTop: -10,
  },
  textStyle: {
    fontSize: 17,
    fontFamily: "Jost-Regular",
  },
});

export default function CustomTextInput({
  errorMessage,
  errorVisible,
  size = "large",
  withClearButton = false,
  onChangeText,
  value,
  secure = false,
  onClear,
  ...otherProps
}) {
  const [secureTextEntry, setSecureTextEntry] = React.useState(secure);

  const toggleSecureEntry = () => {
    setSecureTextEntry(!secureTextEntry);
  };

  const renderSecureIcon = (props) => (
    <TouchableWithoutFeedback testID="secure-icon" onPress={toggleSecureEntry}>
      <Icon {...props} name={secureTextEntry ? "eye-off" : "eye"} />
    </TouchableWithoutFeedback>
  );

  const renderErrorCaption = (errorMessage) => (
    <View style={styles.captionContainer}>
      <Caption
        status="danger"
        testID="error-message"
        style={{ marginTop: 8, marginBottom: 10 }}
      >
        {errorMessage}
      </Caption>
    </View>
  );

  return (
    <>
      <Input
        testID="text-input"
        size={size}
        value={value}
        onChangeText={onChangeText}
        accessoryRight={secure ? renderSecureIcon : null}
        status={errorMessage && errorVisible ? "danger" : "basic"}
        secureTextEntry={secureTextEntry}
        textStyle={styles.textStyle}
        style={{ marginBottom: 10 }}
        {...otherProps}
      />
      {value.length > 0 && withClearButton && (
        <TouchableOpacity testID="clear-icon" onPress={onClear}>
          <MaterialIcons name="clear" size={20} />
        </TouchableOpacity>
      )}
      {errorMessage && errorVisible && renderErrorCaption(errorMessage)}
    </>
  );
}
