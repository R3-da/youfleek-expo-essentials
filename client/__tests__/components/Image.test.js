/* eslint-disable no-undef */
import React from "react";
import { render, cleanup } from "@testing-library/react-native";
import renderer from "react-test-renderer";

import { ApplicationProvider } from "@ui-kitten/components";
import * as eva from "@eva-design/eva";
import Image from "../../app/components/Image";

describe("<Image />", () => {
  afterEach(cleanup);

  it("renders local image correctly", () => {
    const { getByTestId } = render(
      <ApplicationProvider {...eva} theme={eva.light}>
        <Image source={require("../../app/assets/images/icon.png")} />
      </ApplicationProvider>
    );

    expect(() => getByTestId("local-image")).not.toThrow(
      /unable to find an element/i
    );
  });

  it("renders URI image correctly", () => {
    const { getByTestId } = render(
      <ApplicationProvider {...eva} theme={eva.light}>
        <Image imageUri="https://www.sciencemag.org/sites/default/files/styles/article_main_large/public/main_puppies_1280p.jpg?itok=ZZ1podQK" />
      </ApplicationProvider>
    );

    expect(() => getByTestId("uri-image")).not.toThrow(
      /unable to find an element/i
    );
  });

  it("renders and caches URI image correctly", () => {
    const { getByTestId } = render(
      <ApplicationProvider {...eva} theme={eva.light}>
        <Image
          imageUri="https://www.sciencemag.org/sites/default/files/styles/article_main_large/public/main_puppies_1280p.jpg?itok=ZZ1podQK"
          cache={true}
        />
      </ApplicationProvider>
    );

    expect(() => getByTestId("cached-image")).not.toThrow(
      /unable to find an element/i
    );
  });

  it("renders with sharp corners correctly", () => {
    const tree = renderer
      .create(
        <ApplicationProvider {...eva} theme={eva.light}>
          <Image
            imageUri="https://www.sciencemag.org/sites/default/files/styles/article_main_large/public/main_puppies_1280p.jpg?itok=ZZ1podQK"
            rounded={false}
          />
        </ApplicationProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders with round corners correctly", () => {
    const tree = renderer
      .create(
        <ApplicationProvider {...eva} theme={eva.light}>
          <Image imageUri="https://www.sciencemag.org/sites/default/files/styles/article_main_large/public/main_puppies_1280p.jpg?itok=ZZ1podQK" />
        </ApplicationProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
