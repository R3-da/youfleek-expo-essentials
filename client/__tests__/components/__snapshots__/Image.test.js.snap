// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> renders with round corners correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Image
    source={
      Object {
        "uri": "https://www.sciencemag.org/sites/default/files/styles/article_main_large/public/main_puppies_1280p.jpg?itok=ZZ1podQK",
      }
    }
    style={
      Object {
        "borderRadius": 10,
        "height": "100%",
        "width": "100%",
      }
    }
    testID="uri-image"
  />
</View>
`;

exports[`<Image /> renders with sharp corners correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Image
    source={
      Object {
        "uri": "https://www.sciencemag.org/sites/default/files/styles/article_main_large/public/main_puppies_1280p.jpg?itok=ZZ1podQK",
      }
    }
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
    testID="uri-image"
  />
</View>
`;
